src/Ads/AssetSuggestionsService.php

```diff
@@ -227,19 +285,62 @@ protected function get_shop_products( array $args = [] ): array {
 	}
 
 	/**
-	 * Get URL for each attachment using an array of attachment ids.
+	 * Get gallery images ids.
 	 *
-	 * @param array $ids A list of attachments ids.
+	 * @param int $post_id Post ID that contains the gallery.
+	 *
+	 * @return array List of gallery images ids.
+	 */
+	protected function get_gallery_images_ids( int $post_id ): array {
+		$gallery = get_post_gallery( $post_id, false );
+
+		if ( ! $gallery || ! isset( $gallery['ids'] ) ) {
+			return [];
+		}
+
+		return explode( ',', $gallery['ids'] );
+	}
+
+	/**
+	 * Get URL for each attachment using an array of attachment ids and a list of subsizes.
+	 *
+	 * @param array $ids Attachments ids.
+	 * @param array $size_keys Image subsize keys.
 	 *
 	 * @return array A list of attachments urls.
 	 */
-	protected function get_url_attachments_by_ids( array $ids ): array {
+	protected function get_url_attachments_by_ids( array $ids, array $size_keys = [ self::SQUARE_MARKETING_IMAGE_KEY, self::MARKETING_IMAGE_KEY ] ): array {
 		$ids = array_unique( ArrayUtil::remove_empty_values( $ids ) );
+		$ids = array_map( 'intval', $ids );
 
 		$marketing_images = [];
+
 		foreach ( $ids as $id ) {
-			$marketing_images[] = wp_get_attachment_image_url( $id );
+
+			$metadata = wp_get_attachment_metadata( $id );
+
+			foreach ( $size_keys as $size_key ) {
+
+				$minimum_size    = new DimensionUtility( ...$this->image_requirements[ $size_key ]['minimum'] );
+				$recomended_size = new DimensionUtility( ...$this->image_requirements[ $size_key ]['recommended'] );
+				$image_size      = new DimensionUtility( $metadata['width'], $metadata['height'] );
+				$suggested_size  = $this->image_utility->recommend_size( $image_size, $recomended_size, $minimum_size );
+
+				// If the original size matches the suggested size with a precision of +-1px.
+				if ( $suggested_size && wp_fuzzy_number_match( $suggested_size->x, $image_size->x ) && wp_fuzzy_number_match( $suggested_size->y, $image_size->y ) ) {
+					$marketing_images[ $size_key ][] = wp_get_attachment_url( $id, $size_key );
+				} elseif ( isset( $metadata['sizes'][ $size_key ] ) ) {
+					// use the sub size.
```

Comment:
> It's like an image resized to the defined non default size

Correct, the default image sizes are thumbnail, large, medium etc. In our case, we want to create a specific `subsize` for the image assets, so if we need to resize the image, we can add the subsize and `wp_update_image_subsizes` will do the resize for us. 

If the image is suggested again, it will not be needed to create the image again because the subsize already exists.